Build-restructuring, new modules and a compile

codepitbull opened the pull request:
- Build is now a multi module build
- added vertx-lang-scala-stack that contains all currently supported modules
- added vertx-lang-scala-on-the-fly containing a compiler to build code from source-strings and classpath

jponge commented on README.md:
```diff
@@ -1,46 +1,11 @@
+## Structure
+- vertx-lang-scala: The main module containign the actual language implementation
+- vertx-lang-scala-stack: All currently available modules
+- vertx-lang-scala-on-the-fly: A compiler-implementationt that allows to compile Scala-code from source/classpath. 
```
Typo: implementation*t* 😉

codepitbull replied:
fixed

jponge commented:
Why did you remove the `.editorconfig` file? That is always useful.

jponge commented:
`vertx-lang-scala-aggregator` has lost useful metadata on licensing, and it looks like your Maven modules don't have any either. I'd suggest bringing it back.

jponge commented:
In `OnTheFlyCompiler` you should either complete scaladoc, or have none 😉  (e.g. you have empty `@return` / `@param`, etc).

jponge commented:
We have an IP issue regarding `OnTheFlyCompiler` which you took from https://eknet.org/main/dev/runtimecompilescala.html

1. It lacks a license header, and
2. https://eknet.org/main/dev/runtimecompilescala.html does not mention any licensing details except that it was extracted from https://github.com/twitter/util/tree/master/util-eval/src/main/scala/com/twitter/util, and
3. https://github.com/twitter/util/blob/master/util-eval/src/main/scala/com/twitter/util/Eval.scala mentions an Apache v2 License.

Could you please clarify IP provenance and reflect this in the source code?

jponge commented:
You'll probably think that I'm very picky on IP 😉  but all your files (e.g. `ReusableCompiler.scala`) should have a license header.

jponge commented:
Looking at the compilation logs I see recurring warnings. 
Could you please clarify their impact?

```
[INFO] --- scala-maven-plugin:3.2.2:compile (compile) @ vertx-lang-scala_2.11 ---
[WARNING]  Expected all dependencies to require Scala version: 2.11.8
[WARNING]  io.vertx:vertx-lang-scala_2.11:3.4.0-SNAPSHOT requires scala version: 2.11.8
[WARNING]  org.scala-lang.modules:scala-java8-compat_2.11:0.8.0-RC3 requires scala version: 2.11.8
[WARNING]  io.vertx:vertx-lang-scala_2.11:3.4.0-SNAPSHOT requires scala version: 2.11.8
[WARNING]  org.scala-lang:scala-compiler:2.11.8 requires scala version: 2.11.8
[WARNING]  org.scala-lang:scala-reflect:2.11.8 requires scala version: 2.11.8
[WARNING]  org.scala-lang.modules:scala-xml_2.11:1.0.4 requires scala version: 2.11.4
[WARNING] Multiple versions of scala libraries detected!
```

(I don't think it's a big deal between .4 and .8, but still if you can fix the warning it's better)

```
[WARNING] /Users/jponge/Code/vertx/vertx-lang-scala/vertx-lang-scala/src/main/scala/io/vertx/lang/scala/HandlerOps.scala:20: warning: imported `AsyncResult' is permanently hidden by definition of class AsyncResult in package scala
[WARNING] import io.vertx.core.{AsyncResult, Handler}
[WARNING]                       ^
```

There are many of these types of shadowings.

```
[WARNING] /Users/jponge/Code/vertx/vertx-lang-scala/vertx-lang-scala/src/main/scala/io/vertx/lang/scala/ScalaVerticle.scala:52: warning: Adaptation of argument list by inserting () has been deprecated: this is unlikely to be what you want.
[WARNING]         signature: Success.apply[T](value: T): scala.util.Success[T]
[WARNING]   given arguments: <none>
[WARNING]  after adaptation: Success((): Unit)
[WARNING]     startPromise.complete(Success())
```

(deprecations)

```
[WARNING] /Users/jponge/Code/vertx/vertx-lang-scala/vertx-lang-scala/src/test/scala/io/vertx/lang/scala/tck/ApiTest.scala:532: warning: comparing values of types Null and Byte using `==' will always yield false
[WARNING]     nullableTCK.methodWithNullableByteHandler(false,b => assert(null == b))
```

or 

```
[WARNING] /Users/jponge/Code/vertx/vertx-lang-scala/vertx-lang-scala/src/test/scala/io/vertx/scala/codegen/testmodel/TestInterface.scala:301: warning: comparing values of types Int and Null using `==' will always yield false
[WARNING]     if (cached_1 == null) {
[WARNING]                  ^
```

I've seen many of these, this may impact the correctness of your tests.

jponge commented:
That's it for this round, thanks @codepitbull for your work.

codepitbull commented:
Thank you for reviewing!

**The license of OnTheFlyCompiler**
As the original class (before that blogpost) was Apache 2 licensed I added that header (including the Twitter reference) and added references to the original code and the blog-post.
That should fix the licensing issue, at least to my (very limited) understanding.

**Class hiding**
This is a necessary evil and can't be turned off. The classes are actually being hidden as they are the original ones, that's why I reference them with their FQN. Using them with just their classname would hide them and that's what the compiler is telling us here.
The only way to avoid this would be to define custom-type-aliases.
As these won't be used outside the mapped classes and as (at least to me) an additional layer of indirection would reduce readability I decided to keep it the way it is.

**The compiler warnings about the comparisons:**
Those tripped me up when I started but they are a well documented bug:
https://issues.scala-lang.org/browse/SI-602
Sadly, there is no way around these :( 
But they are no problem as the WARNING is simply wrong in these cases.
They also can't be avoided.

codepitbull commented:
**About the deprecations:**
Those come from the mapped Vert.x-code :)
I won't remove them as they are still  available in the Vert.x-API.

jponge commented:
Crystal clear thanks @codepitbull 

Looks good to me!
